---
title: "SSRP: Data Cleanup + Exploration"
author: Nikita Telkar
output: 
  html_document:
    keep_md: yes
  # rmdformats::downcute:
  #   self_contained: true
  #   default_style: "light"
  #   downcute_theme: "default"
    highlight: haddock #tango, pygments, kate, monochrome, espresso, zenburn, haddock, textmate
    toc: true 
    toc_depth: 4
    toc_float: 
      collapsed: true 
      smooth_scroll: false  
    theme: sandstone  #cosmo, paper, lumen, sandstone, simplex, yeti, cerulean, journal, flatly, darkly, readable, spacelab, united

---


***  

### 0.0 Introfilterduction  

Here, we're going to go through what our raw data and clinical data looks like, and then clean or 'wrangle' our data to a useable format. We're also going to make some plots depicting the spread of our data.   

***  

### 1.0 Loading Packages and Data

```{r loading packages}
# rmarkdown packages
library(knitr) 
library(rmarkdown)
library(devtools)

# data wrangling packages
library(tidyverse)
library(reshape2)
library(magrittr)

# data loading packages
library(readxl) 
library(openxlsx)

# plotting/table packages
library(arsenal)
library(kableExtra) 
library(egg)
library(RColorBrewer)
library(ggpubr)
library(emo)

#BiocManager::install(c("limma", "edgeR", "DESeq2", "biomaRt", "miRBaseConverter", "multiMiR", "preprocessCore"))
#remotes::install_github("wvictor14/plomics")

```  

We're going to use a very handy package called the `here` package to navigate between our main folder, and to read in and save files  

```{r here}

library(here)

```

The output should show you which folder the `here` package sets as its root/main directory.  

Now, read in the 2 excel files and the 1 miRNA data file located within your `data` folder

```{r data}

lam_ids <- read_excel(here::here("data", "LamLab_IDs.xlsx"))
rob_ids <- read_excel (here:: here ("data", "RobLab_IDs.xlsx"))
  
mirna <- read_excel(here:: here("data", "miRNA_quantification_raw.xlsx"))

```  

Can you find the function we can use to view the dimensions, meaning the number of rows and columns, of the dataframe? How about to view the structure of the dataframe, meaning the types of variables (e.g., words/characters, numeric, etc)?  

*Hint: there's two functions to view the structure, one is a baseR function, and one is from the tidyverse package.*    

```{r df-structure}

dim(lam_ids)
dim(rob_ids)
dim(mirna)

str(lam_ids)
str(rob_ids) 
str(mirna)

## not sure about the tidyverse one
```


Let's now look at the first few rows of each of the files, or dataframes, that we've loaded and are going to use. I've written down the function using the `lam_ids` datafram, or df, as an example.  

**Do the same for the other dfs**

```{r head}

head(lam_ids)
head(rob_ids)
head(mirna)

```

The `mirna` df shows how expression data usually looks like - your genes are your rows, and your samples are your columns.  

**But here, there's 2 columns for the miRNAs - can you explain why?** 
## The first column is the precursor miRNA which the microRNA before leaving the nucleus and the second column is the mature miRNA which resulted by further manipulations by proteins like Dicer in the cytoplasm.  

***  

### 2.0 Making eDat and pDat  



#### 2.1 pDat

Now, there's two dfs which have some kinds of IDs in them, and it's one ID type that is present as the sample name in the expression df.  

> Having several files containing different aspects of our required data is very common. In most cases, you are the one who will have to go in and clean up the data to fit the format that you prefer

To make one df which contains all of our phenotype information in it, called pDat, we will have to join the two id dfs  

  
But before we join the dfs, R doesn't do well with column/row names which have spaces within the words.  

**So, let's rename the column names which have spaces between the words**    
*Hint: use the `tidyverse::rename` function*. It's best practice to have an underscore `_` as the separator/delimiter between words  

```{r rename-cols}

head(rob_ids)

rob_ids <- rob_ids %>% 
   rename_with(~ gsub(" ","_", .x), contains(" "))   

#I found this function when I was trying to find a way to make this work rename (Case_ID = Case ID); this wouldn't run as there is space between the Case and ID
  

```


Using a function from the tidyverse package:  

**join `lamlab_ids` and `roblab_ids` to create a new dataframe called `pDat`**  

You'll also observe that there's an extra serial number column in the roblab_ids df:  

**remove it too.**  

Further:

**rearrange the order of the columns** to make Case_ID as the first column, followed by ID, lamID, GSC_ID, and then the rest of the columns  
*Hint: specify in your code that you want to use the `select` function from the `dplyr` package*  

```{r making-pdat}

pDat <- cbind(lam_ids, rob_ids) %>% 
  subset(select = - ...1) %>% 
  dplyr:: select(Case_ID, ID, lamID, GSC_ID, everything()) 


```  

Now, check the variable types of each of our variables in pDat  

```{r pdat-class}

str(pDat)

```

We see that all of the variables that should have be numeric are actually encoded as characters - Gestational_Age, Processing_time, Lane, RQS, and totalReadsAligned_miRNA. This will give us problems later on.  

**Convert the above variables from character to numeric**  
*Hint: Use the `mutate_at` function to chnage specific variables*  


```{r pdat-numeric}

pDat <- pDat %>% 
  mutate_at (c("Lane", "Gestational_Age","Processing_time","RQS","totalReadsAligned_miRNA"), as.numeric)

 
```

For the sample which got an `NA` value introduced for its RQS,  

**Replace the NA with 0**  

```{r replace-na}


pDat <- pDat %>% 
  mutate_at(21, ~replace_na(.,0))

```




***  

#### 2.2 eDat

Okay, now if you check our `mirna` df, we can see that the GSC_ID names (the names given to the samples by the GSC (the Genome Sciences Centre) which is the sequencing facility where we've sequenced these samples) are actually not intuitive at all.  

But, now that we've matched the id dataframes, we know exactly which of the GSC_IDs correspond to the Case ID variable.  

**Change the GSC_ID column names to Case_ID names**  

We first have to check whether the GSC_ID order of the column names matches the order of the GSC_ID row names in pDat   

```{r gsc-order}

colnames(mirna)[3:32] == pDat$GSC_ID
# if true for all, proceed. If false, do:

# match(lam_ids$GSC_ID, colnames(mirna)[3:32])
# reorder_idx <- match(pDat$GSC_ID, colnames(mirna)[3:32])
# mirna[3:32] <- mirna[reorder_idx]
# colnames(mirna)[3:32] == pDat$GSC_ID

```

Now, renames the column names in our `mirna` df  
*Hint: use the `colnames` function* to by first selecting the columns to renames, and then the variable you want to rename them by  

```{r renaming-mirna-names}

colnames(mirna)[3:32] <- pDat$Case_ID

```

Now, you know that there's two columns listing miRNA names in the `mirna` df - a lot of our downstream functions require us to only have numeric values in our df, with the gene names as rownames. But, the `column_to_rownames` function only takes the input of one column, but here we have two  

**Combine the precursor and miRNA columns in `mirna` into one column** using `:` as the separator.  

```{r mirna-combine-names}

mirna <- mirna %>% 
  unite("Precursor:miRNA", precursor:miRNA, sep = ":", remove= TRUE)

```

Once combined, 

**Set the new combined column as the rownames for the `mirna` df**  

```{r mirna-rownames}

mirna <- mirna %>%
  column_to_rownames(var="Precursor:miRNA")



```

> This is the format an expression data dataframe should look like 

What is if want to get the average expression for each miRNA across all of the samples?  

**Make a new column showing the average expression across each column per row**  
*Hint: use the `rowSums` function*  

```{r mirna_sums}

mirna <- mirna %>%
  mutate(Sums = rowSums(.))

```

Let's now make our final expression dataframe named `eDat`  

**Make eDat omitting the last expression sum column from the `mirna` df**  

```{r edat}

eDat <- select(mirna, -Sums)


```

***  

### 3.0 Data Exploration  

Okay, now that we have both our `eDat` and `pDat` dfs properly formatted, let's make some plots to display the structure of our dataset  

`ggplot2` (the Grammar of Graphics) package is used to make graphs in R, where it has the following necessary components:  

- `aes` = aesthetics = your x and y axes, and any additional variables you want to deliniate your data by  
- `geoms` = the type to graph you want to plot (e.g., bar plot = geom_bar, scatter plot = geom_point)  

If you look at our eDat df, you'll see that we have each sample as a different column - but if we want to plot our x axis as our samples, ggplot only takes one variable name (column name) for the x axis

And so, we have to first reformat eDat to a df that is plot-able by ggplot. We'll use the slightly advanced function `pivot_longer` to convert our wide eDat df, into a 'longer' df. When this function is used, R forgets to incorporate the rownames into the new df that we make, and hence, we have to change our rownames back into separate columns  

**Convert the rownames of `eDat` into the original two separate columns in a new df**  

```{r rows-to-col}

eDat_nonames <- eDat %>% 
  rownames_to_column(var="Precursor:miRNA") %>% 
  separate("Precursor:miRNA", c("precursor", "miRNA"), ":")


```

Once we have our original format of the expression dataframe (how it looked when we first loaded it in), we'll now convert it to a longer format  

```{r pivot}

eDat_plot <- eDat_nonames %>% 
  pivot_longer(cols = -c(precursor, miRNA), names_to = "Case_ID", values_to = "Reads")

```

You'll see that all of the column names (`names_to`) have been gathered up into one column named Case_ID (to match the name of the column with all the samples in pDat), and all the expression counts (`values_to`) have been formatted into one column I've names as Reads  

To add all the phenotype and expression data together in one df to be able to plot by different variables:  

**Join `eDat_plot` with `pDat`**  

```{r eDat_plot}

eDat_plot <- eDat_plot %>% 
  inner_join( pDat, c ("Case_ID"))

```


Okay, now using pDat, let's make a very simple bar counting the number of samples we have for each 

```{r tri-plot-1}

pDat %>% 
  ggplot(aes(x = Trimester)) +
    #the plus sign is how you add additional arguments/components to your ggplot
  geom_bar(stat = "count") 
    #here, we're not adding a y axis, because we simply want to count the number of samples, and hence specify "count"

```

Let's add some colours to distinguish our samples, and make the width of the bars smaller  

```{r tri-plot-2}

pDat %>% 
  ggplot(aes(x = Trimester, fill = Trimester)) +
  geom_bar(stat = "count", width = 0.6) 



```

Adding labels to the plot, and changing the background:  

```{r tri-plot-3, fig.width=7, fig.height=3}

pDat %>% 
  ggplot(aes(x = Trimester, fill = Trimester)) +
  geom_bar(stat = "count", width = 0.6) +
  theme_minimal() +
  labs(title = "Distribution of Samples by Trimester", subtitle = "n = 30", x = "Trimester", y = "Number of Samples")

```

**Using pDat, I'd like you to plot:**  

- A bar plot showing distribution of samples by sex  

```{r}
pDat %>% 
  ggplot(aes(x = Sex, fill = Sex)) +
  geom_bar(stat = "count", width = 0.6) +
  theme_minimal() +
  labs(title = "Distribution of Samples by Sex", subtitle = "n = 30", x = "Sex", y = "Number of Samples")
```

- A bar plot showing the total reads of aligned miRNAs per sample  


```{r}
pDat %>% 
  ggplot(aes(x = Case_ID, y = totalReadsAligned_miRNA, fill = Case_ID)) +
  geom_bar (stat = "identity", width = 0.6) +
  theme_minimal(base_size = 4) +
  labs(title = "Distribution of Aligned miRNAs per Sample", x = "Case ID", y = "Number of aligned miRNA")
```

- A scatter plot with processing time on the x-axis, and RQS on the y-axis. Is there a correlation between these two variables?  



                    
                    
```{r}
result <-cor.test(x= pDat$RQS,y= pDat$Processing_time, 
                    method = "pearson")

#There is moderate negative correlation of -0.6027559 with 0.0004233 p value. 

pDat %>% 
  ggplot( aes(x=Processing_time, y=RQS)) +
  geom_point() +
  labs(title = "Processing Time vs RQS", subtitle = result$estimate, x = "Processing Time ", y = "RQS")

### Couldn't put this line: estimate: result$estimate
```

**Using eDat_plot, I'd like you to plot:**  

by converting Reads to log2(Reads)  

- A scatter plot showing the expression of hsa-miR-100-3p for each sample, colored by trimester *(Hint: use the `filter` function)*  

```{r}
eDat_plot %>% 
  mutate(Reads=log2(Reads)) %>% 
  filter(miRNA == "hsa-miR-100-3p") %>% 
  ggplot( aes(x=Case_ID, y= Reads, color = Trimester )) +
  geom_point() +
  theme_minimal(base_size = 5) +
  labs(title = "Change of expression of hsa-miR-100-3p in samples over trimesters", x = "Case_ID ", y = "log 2 fold of Reads")

```




- A box plot showing the expressions of hsa-miR-21-5p, hsa-miR-200a-5p, and hsa-miR-373-5p, fill by trimester 
```{r}
eDat_plot %>% 
  mutate(Reads=log2(Reads)) %>% 
  filter (miRNA == c("hsa-miR-21-5p","hsa-miR-200a-5p","hsa-miR-373-5p")) %>% 
  ggplot(aes(x = miRNA, y = Reads, fill = Trimester)) +
  geom_boxplot() +
  theme_minimal(base_size = 12) +
  labs(title = "Distribution of Aligned miRNAs per Sample", x = "Case ID", y = "Number of aligned miRNA")
```


**Make a new variable bucketing the Processing_time variable into 4 categories: <24 hours, 24-48 hours, 48-72 hours, >72 hours**  
*Hint: use the `case_when` function*  

```{r}

eDat_plot <- eDat_plot %>% 
  mutate(hours = case_when(
    Processing_time <= 24 ~ "<24 hours",
    between(Processing_time, 24, 48) ~ "24-48 hours", 
    between(Processing_time, 49, 72) ~ "48-72 hours", 
    Processing_time >= 73 ~ ">72 hours",
  ))



```

- Count the number of samples in each of the above formed categories

```{r}
eDat_plot %>% 
  count (hours) 

```

- A scatter plot showing the expression of hsa-miR-520c-3p per sample for each category, colour by trimester,  `+ facet_wrap(~Trimester)`

```{r}
eDat_plot %>% 
  mutate(Reads=log2(Reads)) %>% 
  filter(miRNA == "hsa-miR-520c-3p") %>% 
  ggplot( aes(x = hours, y = Reads, color = Trimester )) +
  geom_point() +
  facet_wrap(~Trimester) +
  theme_minimal(base_size = 7) 

  
```


***

### 4.0 Saving files  

**Save `eDat` and `pDat` as `.RDS` objects within your `data` folder

```{r saving}
saveRDS(eDat, here::here ("data", "eDat.RDS"))
saveRDS(pDat, here::here ("data", "pDat.RDS"))

```

It was super fun!
***  

